### PR TITLE
Support claim attribute mappings in bulk

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Related issue:
- https://github.com/wso2/product-is/issues/12253

Add new claimMetadataManagementService methods to support updating claim attribute mappings in bulk.

### CODE REVIEW

**Review by JKAUSHALYA — CHANGES_REQUESTED** [review]

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java [hunk 1]
```diff
@@ -92,6 +92,17 @@ void renameClaimDialect(ClaimDialect oldClaimDialect, ClaimDialect newClaimDiale
...
 
+    /**
+     * Update attribute mappings for local claims in bulk
+     *
+     * @param localClaimList
+     * @param tenantDomain
+     * @param userStoreDomain
+     * @throws ClaimMetadataException
+     */
+    void updateLocalClaimMappings(List<LocalClaim> localClaimList, String tenantDomain, String userStoreDomain)
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java]: Addressed with commit c0667c87a60e5998c84fe0a979af482fd57b6b15

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java [hunk 2]
```diff
@@ -176,4 +187,14 @@ default String getMaskingRegexForLocalClaim(String localClaimURI, String tenantD
...
     }
 
+    /**
+     * To validate claim mapping attributes of local claims.
+     *
+     * @param localClaimList List of Local claims
+     * @param tenantDomain  Tenant domain
+     * @throws ClaimMetadataException If an error occurred while validating local claims.
+     */
+    public void validateClaimAttributeMapping(List<LocalClaim> localClaimList,  String tenantDomain)
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java]: Addressed with commit c0667c87a60e5998c84fe0a979af482fd57b6b15

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java]: Addressed with commit c0667c87a60e5998c84fe0a979af482fd57b6b15

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java [hunk 3]
```diff
@@ -229,6 +230,23 @@ public void updateLocalClaim(LocalClaim localClaim, String tenantDomain) throws
...
     }
 
+    @Override
+    public void updateLocalClaimMappings(List<LocalClaim> localClaimList, String tenantDomain, String userStoreDomain)
+            throws ClaimMetadataException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+        for (LocalClaim localClaim : localClaimList) {
+            ClaimMetadataEventPublisherProxy.getInstance().publishPreUpdateLocalClaim(tenantId, localClaim);
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java]: Addressed with commit c0667c87a60e5998c84fe0a979af482fd57b6b15

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java [hunk 4]
```diff
@@ -221,6 +221,52 @@ public void updateLocalClaim(LocalClaim localClaim, int tenantId) throws ClaimMe
...
+     * @param localClaimList  list of local claims
+     * @param tenantId        Tenant Id
+     * @param userStoreDomain Domain name
+     * @throws ClaimMetadataException If an error occurred while updating local claims
+     */
+    public void updateLocalClaimMappings(List<LocalClaim> localClaimList, int tenantId, String userStoreDomain)
+            throws ClaimMetadataException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        PreparedStatement prepStmt = null;
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java]: Addressed with commit c0667c87a60e5998c84fe0a979af482fd57b6b15

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java [hunk 5]
```diff
@@ -229,6 +230,31 @@ public void updateLocalClaim(LocalClaim localClaim, String tenantDomain) throws
...
+        ClaimMetadataEventPublisherProxy claimMetadataEventPublisherProxy =
+                ClaimMetadataEventPublisherProxy.getInstance();
+        for (LocalClaim localClaim : localClaimList) {
+            claimMetadataEventPublisherProxy.publishPreUpdateLocalClaim(tenantId, localClaim);
+        }
+
+        this.localClaimDAO.updateLocalClaimMappings(localClaimList, tenantId, userStoreDomain);
+
+        for (LocalClaim localClaim : localClaimList) {
+            ClaimMetadataEventPublisherProxy.getInstance().publishPostUpdateLocalClaim(tenantId, localClaim);
```
Comment (JKAUSHALYA) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java:254]: We can use the same local variable here as well.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java [hunk 6]
```diff
@@ -221,6 +221,51 @@ public void updateLocalClaim(LocalClaim localClaim, int tenantId) throws ClaimMe
...
+     * @param userStoreDomain Domain name
+     * @throws ClaimMetadataException If an error occurred while updating local claims
+     */
+    public void updateLocalClaimMappings(List<LocalClaim> localClaimList, int tenantId, String userStoreDomain)
+            throws ClaimMetadataException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+
+        try {
+            // Start transaction
```
Comment (JKAUSHALYA) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java:238]: Correct commenting format is // Comment.  
A Comment always needs to be end with a full stop.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java (hunk 5 above)
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementServiceImpl.java]: Addressed with commit abcc28ca2bde2f2eea597841ec5000f0cddd7202

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java (hunk 6 above)
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java]: Addressed with commit abcc28ca2bde2f2eea597841ec5000f0cddd7202

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java [hunk 7]
```diff
@@ -162,6 +171,29 @@ default void removeAllClaims(int tenantId) throws ClaimMetadataException {
 
     }
 
+    /**
+     * Update attribute mappings for local claims in bulk.
+     *
+     * @param localClaimList    list of local claims.
+     * @param tenantDomain      tenant domain name.
+     * @param userStoreDomain   user store domain name.
+     * @throws ClaimMetadataException if error occurs while updating local claims.
```
Comment (ashendes) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:180]: Descriptions should be in sentence case.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java [hunk 8]
```diff
@@ -162,6 +171,29 @@ default void removeAllClaims(int tenantId) throws ClaimMetadataException {
...
+    default void updateLocalClaimMappings(List<LocalClaim> localClaimList, String tenantDomain, String userStoreDomain)
+            throws ClaimMetadataException {
+    }
+
+    /**
+     * To validate claim mapping attributes of local claims.
+     *
+     * @param localClaimList List of Local claims
+     * @param tenantDomain  Tenant domain
+     * @throws ClaimMetadataException If an error occurred while validating local claims.
```
Comment (ashendes) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:191]: Formatting issues. Inconsistent spacing between param name and descriptions.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java [hunk 9]
```diff
@@ -75,6 +75,21 @@ public void updateLocalClaim(LocalClaim localClaim, int tenantId) throws ClaimMe
         localClaimInvalidationCache.clearCacheEntry(tenantId, tenantId);
     }
 
+    /**
+     * Update attribute claim mappings related to tenant id and domain in bulk.
+     *
+     * @param localClaimList  list of local claims.
```
Comment (ashendes) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java:81]: Description should be in sentence case.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java [hunk 10]
```diff
@@ -221,6 +221,51 @@ public void updateLocalClaim(LocalClaim localClaim, int tenantId) throws ClaimMe
         }
     }
 
+    /**
+     * Update attribute claim mappings related to tenant id and domain.
+     *
+     * @param localClaimList  list of local claims.
```
Comment (ashendes) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java:227]: Fix case.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java (hunk 7 above)
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:180]: Addressed with commit 2e3f3c951772d195bcbedb1a7b81f4815b531259 in https://github.com/wso2/carbon-identity-framework/pull/3774 PR

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java (hunk 8 above)
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/ClaimMetadataManagementService.java:191]: Addressed with commit 2e3f3c951772d195bcbedb1a7b81f4815b531259 in https://github.com/wso2/carbon-identity-framework/pull/3774 PR

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java (hunk 9 above)
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/CacheBackedLocalClaimDAO.java:81]: Addressed with commit 2e3f3c951772d195bcbedb1a7b81f4815b531259 in https://github.com/wso2/carbon-identity-framework/pull/3774 PR

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java (hunk 10 above)
Comment (Thisara-Welmilla) [reply components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAO.java:227]: Addressed with commit 2e3f3c951772d195bcbedb1a7b81f4815b531259 in https://github.com/wso2/carbon-identity-framework/pull/3774 PR

**Pull request merged**